docs/jax-101/03-vectorization.ipynb

```diff
@@ -0,0 +1,376 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "id": "5Obro91lwE_s",
+    "outputId": "061983c6-2faa-4a54-83a5-d2a823f61087"
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "DeviceArray([11., 20., 29.], dtype=float32)"
+      ]
+     },
+     "execution_count": 1,
+     "metadata": {
+      "tags": []
+     },
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import jax\n",
+    "import jax.numpy as jnp\n",
+    "\n",
+    "x = jnp.arange(5)\n",
+    "w = jnp.array([2., 3., 4.])\n",
+    "\n",
+    "def convolve(x, w):\n",
+    "  output = []\n",
+    "  for i in range(1, len(x)-1):\n",
+    "    output.append(jnp.dot(x[i-1:i+2], w))\n",
+    "  return jnp.array(output)\n",
+    "\n",
+    "convolve(x, w)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "z_nPhEhLRysk"
+   },
+   "source": [
+    "Suppose we would like to apply this function to a batch of weights `w` to a batch of vectors `x`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "id": "rHQJnnrVUbxE"
+   },
+   "outputs": [],
+   "source": [
+    "xs = jnp.stack([x, x])\n",
+    "ws = jnp.stack([w, w])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "ghaJQW1aUfPi"
+   },
+   "source": [
+    "The most naive option would be to simply loop over the batch in Python:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {
+    "id": "yM-IycdlzGyJ",
+    "outputId": "07ed6ffc-0265-45ef-d585-4b5fa7d221f1"
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "DeviceArray([[11., 20., 29.],\n",
+       "             [11., 20., 29.]], dtype=float32)"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {
+      "tags": []
+     },
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "def manually_batched_convolve(xs, ws):\n",
+    "  output = []\n",
+    "  for i in range(xs.shape[0]):\n",
+    "    output.append(convolve(xs[i], ws[i]))\n",
+    "  return jnp.stack(output)\n",
+    "\n",
+    "manually_batched_convolve(xs, ws)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "VTh0l_1SUlh4"
+   },
+   "source": [
+    "This produces the correct result, however it is not very efficient.\n",
+    "\n",
+    "In order to batch the computation efficiently, you would normally have to rewrite the function manually to ensure it is done in vectorised form. This is not particularly difficult to implement, but does involve changing how the function treats indices, axes, and other parts of the input.\n",
```

Comment:
Suggestion: nit 🇬🇧 vs 🇺🇸 "vectorized" (I prefer "vectorised" 😎 though but the tech style guides recommend 🇺🇸 English 🤷 )

```suggestion
    "In order to batch the computation efficiently, you would normally have to rewrite the function manually to ensure it is done in vectorized form. This is not particularly difficult to implement, but does involve changing how the function treats indices, axes, and other parts of the input.\n",
```